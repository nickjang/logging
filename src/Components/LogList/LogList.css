.log-list {
  background-color: white;
  width: fit-content;
  margin: 2rem auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.log-list h3 {
  margin: 2rem 0px;
  text-align: center;
}

.log-list p {
  margin: 0;
}

.log-list .log-list-form select {
  width: 100%;
  min-height: 3rem;
  height: max-content;
  max-height: 30rem;
  border: 1px solid white;
  border-top: 1px solid lightgray;
  border-bottom: 1px solid lightgray;
  border-radius: 2px;
  overflow-x: scroll;
}

.log-list .log-list-form option {
  font-family: var(--timestamp-font);
  font-weight: 700;
  font-size: 20px;
  text-align: left;
  padding: 12px;
}

.log-list .note {
  text-align: center;
}
