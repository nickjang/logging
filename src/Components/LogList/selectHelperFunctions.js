
/**
 * Allow dragging over item to select or 
 * unselect it. 
 * If mouse is down, then select/unselect
 * based on the first option interacted with.
 */
const handleMouseOver = (id) => {
  if (this.state.mouseDown) {
    const isSelected = this.state.selected.includes(id);
    let selected = [...this.state.selected];
    // If firstInteracted doesn't exist, select/unselect option and set it as firstInteracted
    if (!this.state.firstInteracted.id) {
      if (isSelected) selected = selected.filter(log => log !== id);
      else selected.push(id);
      this.setState({ firstInteracted: { id, isSelected: !isSelected }, selected });
    } else {
      // If firstInteracted exists, select based on it
      if (this.state.firstInteracted.isSelected) {
        if (isSelected) return;
        this.setState({ selected: [...selected, id] });
      } else {
        if (!isSelected) return;
        selected = selected.filter(log => log !== id);
        this.setState({ selected });
      }
    }
  }
}

/**
 * All select/unselect multiple logs with shift.
 * If shift key down and firstInteracted exists, 
 * select/unselect indexes from firstInteracted
 * to log clicked based on firstInteracted. 
 * If shift key down, then select/unselect. 
 * If shift not key down, select/unselect item 
 * that was mouseupped. 
 */
const handleMouseUp = (id) => {
  const isSelected = this.state.selected.includes(id);
  let selected = [...this.state.selected];

  if (this.state.shiftPressing) {
    // if firstInteracted exists, select based on firstInteracted
    if (this.state.firstInteracted.id) {
      if (this.state.firstInteracted.isSelected) {
        if (isSelected) return;
        this.setState({ selected: [...selected, id] });
      } else {
        if (!isSelected) return;
        selected = selected.filter(log => log !== id);
        this.setState({ selected });
      }
      // if firstInteracted doesn't exist, select/unselect option and set it as firstInteracted
    } else {
      if (isSelected) selected = selected.filter(log => log !== id);
      else selected.push(id);
      this.setState({ firstInteracted: { id, isSelected: !isSelected }, selected })
    }
  } else {
    if (isSelected) selected = selected.filter(log => log !== id);
    else selected.push(id);
    this.setState({ selected });
  }
}

const handleShiftKeyDown = (e) => {
  if (e.keyCode === 16) this.setState({ shiftPressing: true });
}

const handleShiftKeyUp = (e) => {
  if (e.keyCode === 16) this.setState({ shiftPressing: false, firstInteracted: '' });
}

const updateSelectedLogs = (logs) => {
}

/*
componentDidMount = () => {
  document.addEventListener('keydown', (e) => this.handleShiftKeyDown(e));
  document.addEventListener('keyup', (e) => this.handleShiftKeyUp(e));
  document.addEventListener('mousedown', () => this.setState({ mouseDown: true }));
  document.addEventListener('mouseup', () => this.setState({ mouseDown: false, firstInteracted: '' }));
}

<option>
onMouseOver={(e) => this.handleMouseOver(e.target.value)}
onMouseUp={(e) => this.handleMouseUp(e.target.value)}

state:
firstInteracted: {
  id: '',
  isSelected: false
}
shiftPressing: false,
mouseDown: false
*/