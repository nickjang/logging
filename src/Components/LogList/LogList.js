import React, { Component } from 'react';
import PropTypes from 'prop-types';
import formatLog from '../Formatter/formatLog';
import './LogList.css';

class LogList extends Component {
  state = {
    selected: [],
    firstInteracted: {
      id: '',
      isSelected: false
    },
    shiftPressing: false,
    mouseDown: false
  }

  //flatten picked before getting logs
  //do that before here so loglist doesn't flatten every rerender
  //logs should be chronological, even from different projects
  //select logs based on log's start time

  /**
   * Allow dragging over item to select or 
   * unselect it. 
   * If mouse is down, then select/unselect
   * based on the first option interacted with.
   */
  handleMouseOver = (id) => {
    if (this.state.mouseDown) {
      const isSelected = this.state.selected.includes(id);
      let selected = [...this.state.selected];
      // If firstInteracted doesn't exist, select/unselect option and set it as firstInteracted
      if (!this.state.firstInteracted.id) {
        if (isSelected) selected = selected.filter(log => log !== id);
        else selected.push(id);
        this.setState({ firstInteracted: { id, isSelected: !isSelected }, selected });
      } else {
        // If firstInteracted exists, select based on it
        if (this.state.firstInteracted.isSelected) {
          if (isSelected) return;
          this.setState({ selected: [...selected, id] });
        } else {
          if (!isSelected) return;
          selected = selected.filter(log => log !== id);
          this.setState({ selected });
        }
      }
    }
  }

  /**
   * All select/unselect multiple logs with shift.
   * If shift key down and firstInteracted exists, 
   * select/unselect indexes from firstInteracted
   * to log clicked based on firstInteracted. 
   * If shift key down, then select/unselect. 
   * If shift not key down, select/unselect item 
   * that was mouseupped. 
   */
  handleMouseUp = (id) => {
    const isSelected = this.state.selected.includes(id);
    let selected = [...this.state.selected];

    if (this.state.shiftPressing) {
      // if firstInteracted exists, select based on firstInteracted
      if (this.state.firstInteracted.id) {
        if (this.state.firstInteracted.isSelected) {
          if (isSelected) return;
          this.setState({ selected: [...selected, id] });
        } else {
          if (!isSelected) return;
          selected = selected.filter(log => log !== id);
          this.setState({ selected });
        }
        // if firstInteracted doesn't exist, select/unselect option and set it as firstInteracted
      } else {
        if (isSelected) selected = selected.filter(log => log !== id);
        else selected.push(id);
        this.setState({ firstInteracted: { id, isSelected: !isSelected }, selected })
      }
    } else {
      if (isSelected) selected = selected.filter(log => log !== id);
      else selected.push(id);
      this.setState({ selected });
    }
  }

  handleShiftKeyDown = (e) => {
    if (e.keyCode === 16) this.setState({ shiftPressing: true });
  }

  handleShiftKeyUp = (e) => {
    if (e.keyCode === 16) this.setState({ shiftPressing: false, firstInteracted: '' });
  }

  updateSelectedLogs = (logs) => {
  }

  componentDidMount = () => {
    document.addEventListener('keydown', (e) => this.handleShiftKeyDown(e));
    document.addEventListener('keyup', (e) => this.handleShiftKeyUp(e));
    document.addEventListener('mousedown', () => this.setState({ mouseDown: true }));
    document.addEventListener('mouseup', () => this.setState({ mouseDown: false, firstInteracted: '' }));
  }

  render() {
    const options = this.props.logs.map(log =>
      <option
        value={log.id}
        onMouseOver={(e) => this.handleMouseOver(e.target.value)}
        onMouseUp={(e) => this.handleMouseUp(e.target.value)}
      >{formatLog(log.value, log.format)}
      </option>
    );

    return (
      <section className='logs'>
        <h3>Logs</h3>
        <span className='form-status'>{this.props.status || 'Please select logs to view from the side bar.'}</span>
        <form action='' id='logs-select-form'>
          {/* only accepts option or optgroup */}
          <select
            id='logs-list-select'
            name='logs'
            multiple
            value={this.state.selected}
            aria-label='Select logs you would like to format. Selected logs can be formatted in any format logs box on the page.'
            aria-required='true'
            onChange={(e) => this.updateSelectedLogs(e.target.value)}
          >
            {options}
          </select>
          <div className='curly-brace'>
            <img src=''
              alt='The top of a curly brace that show which logs are selected. Selected logs can be formatted in any format logs box on the page.' />
            <img src=''
              alt='The middle of a curly brace that shows which logs are selected. Selected logs can be formatted in any format logs box on the page.' />
            <img src=''
              alt='The bottom of a curly brace that shows which logs are selected. Selected logs can be formatted in any format logs box on the page.' />
          </div>
        </form>
        <aside className='note'>Logs are automatically updated to their project.</aside>
      </section>
    );
  }
}

LogList.defaultProps = {
  logs: [],
  status: ''
}

LogList.propTypes = {
  logs: PropTypes.array,
  status: PropTypes.string
}

export default LogList;