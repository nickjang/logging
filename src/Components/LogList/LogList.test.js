import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import LogList from './LogList';

describe(`LogList component`, () => {
  it('renders LogList by default', () => {
    const wrapper = shallow(<LogList />);
    expect(toJson(wrapper)).toMatchSnapshot()
  });
})