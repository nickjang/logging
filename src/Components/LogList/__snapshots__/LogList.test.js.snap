// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogList component renders LogList by default 1`] = `
<section
  className="log-list"
>
  <p
    className="note"
  >
    Select logs to format them. The original time is preserved.
  </p>
  <p
    className="note"
  >
    Try 01:01 to get the original.
  </p>
  <h3
    className="lg-title"
  >
    Logs
  </h3>
  <form
    action=""
    className="log-list-form mt-1"
    id="logs-select-form"
  >
    <select
      aria-label="Select logs you would like to format. Selected logs can be formatted in any format logs box on the page."
      aria-required="true"
      id="logs-list-select"
      multiple={true}
      name="logs"
      onChange={[Function]}
    />
  </form>
</section>
`;
